# Assignment:

For this task, you will the [Boston Housing Data](https://drive.google.com/file/d/1EtNMLkPCltdm8TH_HA-c8D4L75apgzkn/view?usp=sharing) set that you used in a previous exercise with linear regression.

# Preliminary steps

In [ ]:
# import libraries
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeRegressor

In [ ]:
# load data
path = '/content/drive/MyDrive/Coding Dojo/06 Week 6: Regression Models/Boston_Housing_from_Sklearn.csv'
df = pd.read_csv(path)

In [ ]:
# explore data
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 7 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   CRIM     506 non-null    float64
 1   NOX      506 non-null    float64
 2   RM       506 non-null    float64
 3   AGE      506 non-null    float64
 4   PTRATIO  506 non-null    float64
 5   LSTAT    506 non-null    float64
 6   PRICE    506 non-null    float64
dtypes: float64(7)
memory usage: 27.8 KB


In [ ]:
# explore data
df.sample(10)

CRIM    NOX     RM   AGE  PTRATIO  LSTAT  PRICE
300  0.04417  0.400  6.871  47.4     14.8   6.07   24.8
0    0.00632  0.538  6.575  65.2     15.3   4.98   24.0
88   0.05660  0.489  7.007  86.3     17.8   5.50   23.6
471  4.03841  0.532  6.229  90.7     20.2  12.87   19.6
239  0.09252  0.428  6.606  42.2     16.6   7.37   23.3
328  0.06617  0.460  5.868  25.8     16.9   9.97   19.3
71   0.15876  0.413  5.961  17.5     19.2   9.88   21.7
250  0.14030  0.431  6.487  13.0     19.1   5.90   24.4
40   0.03359  0.428  7.024  15.8     18.3   1.98   34.9
47   0.22927  0.448  6.030  85.5     17.9  18.80   16.6

In [ ]:
# check for duplicates
df.duplicated().sum()

0

In [ ]:
# check for missing values
df.isna().sum().sum()

0

In [ ]:
# assign target vector and features matrix
target = 'PRICE'
y = df[target]
X = df.drop(columns = target)

In [ ]:
# train test split (model validation)
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state = 42)

# 1. Run a regression tree model with default parameters (unlimited depth).



In [ ]:
# instantiate decision tree model with all default parameters
dec_tree = DecisionTreeRegressor(random_state = 42)

In [ ]:
# fit decision tree model on training data only
dec_tree.fit(X_train, y_train)

DecisionTreeRegressor(random_state=42)

# 2. Determine the depth of the default tree.



In [ ]:
dec_tree.get_depth()

20

# 3. Try different values for max_depth and determine the optimal value based on the best (highest) r2 value. What is the optimal max_depth based on your trials?



In [ ]:
# tune hyperparameter (max_depth) of tree to find highest r2 value

# set list of each possible depth between 2 (lowest depth) and 20 (default tree depth)
depths = list(range(2, 20))

# create dataframe to store number of levels of each tree (depth, as index) with that tree's test data score and train data score (r2 values)
scores = pd.DataFrame(index = depths, columns = ['Test Score', 'Train Score'])

# for each depth
for depth in depths:

  # make a decision tree with specified depth
  dec_tree = DecisionTreeRegressor(max_depth = depth, random_state = 42)

  # fit the tree on the training data only (X and y to learn relationship between them)
  dec_tree.fit(X_train, y_train)

  # assign train_score to the r2 value (.score()) of the tree predicting the training data
  train_score = dec_tree.score(X_train, y_train)

  # assign test_score to the r2 value (.score()) of the tree predicting the testing data
  test_score = dec_tree.score(X_test, y_test)

  # put train_score in the scores df
  scores.loc[depth, 'Train Score'] = train_score

  # put test_score in the scores df
  scores.loc[depth, 'Test Score'] = test_score

In [ ]:
# sort the scores df on test score to find the highest r2 score on the testing data
sorted_scores = scores.sort_values(by ='Test Score', ascending = False)
sorted_scores.head()

Test Score Train Score
7    0.846377    0.958517
10    0.84601    0.986796
11   0.829736      0.9911
12   0.827102    0.995358
6    0.825985    0.942742

The optimal max_depth for this model is 7.

# 4. What is the r2 of your final model on the training set and on the test set?



The R^2 score of this decision tree model with a max_depth of 7 on the training data is about 0.96, and on the testing data it is about 0.85.